### PR TITLE
fix: handle exceptions when using /0 as a URL

### PR DESCRIPTION
**Pull request by wilr**
Fix for https://github.com/silverstripe/silverstripe-cms/issues/2680

### CODE REVIEW

**Comment by michalkleiner** [comment]
Thanks Will. Since this is fixing a bug, it can target `4.12`.
As per @GuySartorelli's [comment](https://github.com/silverstripe/silverstripe-cms/pull/2685#issuecomment-968432261), the fix may not be as simple as this, so unit test(s) would be needed as well.

**Comment by shanholmes** [comment]
This is still an issue and has since carried over to the 5 branch of the CMS.

**Comment by GuySartorelli** [comment]
Rebased, added tests, and made changes necessary for those tests to pass.

File: tests/php/Model/SiteTreeTest.php [hunk 1]
```diff
@@ -459,6 +460,7 @@ public function testGetByLink()
         $about    = $this->objFromFixture('Page', 'about');
         $staff    = $this->objFromFixture('Page', 'staff');
         $product  = $this->objFromFixture('Page', 'product1');
+        $numeric0  = $this->objFromFixture('Page', 'numeric0');
```
Comment (michalkleiner) [inline tests/php/Model/SiteTreeTest.php]: nitpick
```suggestion
        $numeric0 = $this->objFromFixture('Page', 'numeric0');
```

File: code/Model/SiteTree.php [hunk 2]
```diff
@@ -1633,6 +1633,11 @@ public function requireDefaultRecords()
         }
     }
 
+    private function hasURLSegment(): bool
+    {
+        return $this->URLSegment !== null && $this->URLSegment !== '';
```
Comment (michalkleiner) [inline code/Model/SiteTree.php]: I guess excluding `false` is not necessary here, but would qualify as having a URL segment.

File: tests/php/Model/SiteTreeTest.php (hunk 1 above)
Comment (ChloeSartorelli) [reply tests/php/Model/SiteTreeTest.php]: Done

File: tests/php/Model/SiteTreeTest.php [hunk 3]
```diff
@@ -489,26 +493,30 @@ public function testGetByLinkAbsolute()
...
```
Comment (michalkleiner) [inline tests/php/Model/SiteTreeTest.php]: nitpick
```suggestion
        $numeric0 = $this->objFromFixture('Page', 'numeric0');
```

File: tests/php/Model/SiteTreeTest.php [hunk 4]
```diff
@@ -489,26 +493,30 @@ public function testGetByLinkAbsolute()
...
         $this->assertEquals($staff->ID, SiteTree::get_by_link(Controller::join_links($base, $staff->Link()), false)->ID);
         $this->assertEquals($product->ID, SiteTree::get_by_link(Controller::join_links($base, $product->Link()), false)->ID);
+        $this->assertEquals($numeric0->ID, SiteTree::get_by_link(Controller::join_links($base, $numeric0->Link()), false)->ID);
     }
 
     public function testRelativeLink()
     {
         $about    = $this->objFromFixture('Page', 'about');
         $staff    = $this->objFromFixture('Page', 'staff');
+        $numeric0    = $this->objFromFixture('Page', 'numeric0');
```
Comment (michalkleiner) [inline tests/php/Model/SiteTreeTest.php]: cs
```suggestion
        $numeric0 = $this->objFromFixture('Page', 'numeric0');
```

**Review by michalkleiner — APPROVED** [review]

File: code/Model/SiteTree.php [hunk 5]
```diff
@@ -456,7 +456,7 @@ public static function get_by_link($link, $cache = true)
         $parentIDExpr = sprintf('"%s"."ParentID"', $tableName);
 
         $link = trim(Director::makeRelative($link) ?? '', '/');
-        if (!$link) {
+        if ($link === false || $link === null || $link === '') {
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:459]: Note that while `trim()` will only ever return a string, having the full check here avoids issues if we ever remove that trim.

File: code/Model/SiteTree.php (hunk 2 above)
Comment (ChloeSartorelli) [reply code/Model/SiteTree.php]: I don't think it's necessary since it's unlikely to ever be `false` - but it doesn't hurt to add to avoid that edge case.
Added.

File: tests/php/Model/SiteTreeTest.php (hunk 3 above)
Comment (ChloeSartorelli) [reply tests/php/Model/SiteTreeTest.php]: Done

File: tests/php/Model/SiteTreeTest.php (hunk 4 above)
Comment (ChloeSartorelli) [reply tests/php/Model/SiteTreeTest.php]: Done

**Review by emteknetnz — APPROVED** [review]
Tested locally, works good
